docs/design/active-active/active-active.md

```diff
@@ -0,0 +1,306 @@
+
+1. Pick the cluster that received the start workflow request
+   This is the default behavior and requires no additional configuration.
+
+2. Pick the cluster based on an external entity specified in the start workflow request.
+   This is useful when the user wants to dynamically change which cluster a workflow should be active in during the lifetime of workflow.
+   e.g. In "Domain-AA-User-Location" example, a workflow that is started in cluster-0 but user moved to another region and the workflow needs to be active in cluster-1 going forward.
+
+
+Workflow start parameters:
+
+| Parameter | Description | Example |
+|-----------|-------------|---------|
+| active-region.lookup-source | Where to look up to determine the active region of the workflow for a given key. Ideally it should support the watch API. A corresponding watcher must be implemented within Cadence. | "custom.user-location-provider" |
+| active-region.lookup-key | The key to look up from the above source. | "userid-123" |
+
+The lookup source "custom.user-location-provider" is a custom provider that can be implemented by the user. Initially we are not going to define an RPC protocol for this and it has to be implemented within Cadence.
+
+
+Workflow start request determines which cluster selection strategy to be used.
+
+| Has active-region.lookup-key | Has active-region.origin | Strategy |
+|------------------------------|-------------------------|----------|
+| No                           | No                      | Type 1   |
+| Yes                          | No                      | Type 1   |
+| No                           | Yes                     | Type 2   |
+
+If none of the new parameters are set, the default behavior is to pick the cluster that received the start workflow request.
+It's not allowed to have both active-region.lookup-key and active-region.origin set.
+
+
+### EntityActiveRegion Lookup Table
+- Primary key of the table is <EntityType, EntityKey>
+- A new plugin system will be introduced to watch supported entity types and populate the table
+- Watchers are per cluster group
+    - There’s one primary cluster for the group and watcher runs on that cluster
+- This table is replicated to all clusters in the cluster group from the primary cluster
+- Failover Version
+    - Each entity record has a failover version to be used as workflow task versions
+    - Initially set based on cluster metadata
+    - Incremented based on Increment value from cluster metadata whenever mapping for a key changes
+
+
+| EntityType | EntityKey | Region | Failover Version | LastUpdated |
+|------------|-----------|--------|------------------|-------------|
+| user-location | seattle   | us-west    | 1                |             |
```

Comment:
When and how are these records inserted?

Reply:
It's briefly explained in the bullet points above. For each such 3rd party entity type there needs to be a watcher implemented which populates this table. e.g. user-location entity type would be managed by a new UserLocationWatcher which runs in primary cluster as a global singleton